### PR TITLE
PP-5402-Modify domain objects and JSON serialization/deserialization classes-2

### PR DESCRIPTION
**Pull request by nimalank7**
## WHAT YOU DID

This PR modifies the TelephoneChargeRequest, TelephoneChargeResponse, ChargeEntity, CardDetailsEntity, PaymentOutcome and State by adding constructors, setters and changing from primitive type integer to the object type of Long.

This also modifies the TelephonePaymentJSONTest to pass in Long instead of int.

We can map amount, reference, description, status, email and cardDetails from the JSON into the ChargeEntity fields so we will need a constructor to take these. Any other remaining fields in the JSON will have to go into ExternalMetadata.

For constructing the TelephonePaymentResponse we need 2 constructors for State. This is because for success we don't want to return a code but with failure we do.

Since the Supplemental object in the JSON is optional then we need 2 constructors in PaymentOutcome to handle this (e.g. one that takes it one that doesn't). Since status in PaymentOutcome is always required in the JSON and since code won't be provided with status success we need a PaymentOutcome constructor that just takes in status.

Please review https://github.com/alphagov/pay-connector/pull/1610 before merging this PR in.

### CODE REVIEW

**Comment by nimalank7** [comment]
Rebased from updated master.

File: src/main/java/uk/gov/pay/connector/charge/model/CardDetailsEntity.java [hunk 1]
```diff
@@ -43,6 +43,15 @@
 
     public CardDetailsEntity() {
     }
+    
+    // For telephone payments
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/connector/charge/model/CardDetailsEntity.java:47]: I wouldn't bother with this sort of comment - it doesn't add to any understanding, and is likely to become stale - maybe some other thing will be built that also needs this constructor in the future

File: src/main/java/uk/gov/pay/connector/charge/model/telephone/State.java [hunk 2]
```diff
@@ -13,6 +13,12 @@
     public State() {
         // For Jackson serialization
     }
+    
+    public State(String status, Boolean finished, String message) {
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/connector/charge/model/telephone/State.java:17]: it would be good to understand why these changes are necessary - your PR describes what you have done, but not the rationale, which makes it hard to evaluate some of your choices

**Review by mrlumbu — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/pay/connector/charge/model/CardDetailsEntity.java (hunk 1 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/connector/charge/model/CardDetailsEntity.java:47]: Sure I'll remove it.

File: src/main/java/uk/gov/pay/connector/charge/model/telephone/State.java (hunk 2 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/connector/charge/model/telephone/State.java:17]: Apologies - for constructing the TelephonePaymentResponse I'll need a constructor which have these parameters. Similarly for persisting to the database I'll need to be able to store stuff from the JSON into the ExternalMetadata. I thought it'd be easier to just group these changes together. I'll update the PR description accordingly.

Comment (nimalank7) [reply src/main/java/uk/gov/pay/connector/charge/model/telephone/State.java:17]: Yes - I've updated the PR description accordingly. The constructors are required for certain fields in the JSON being optional. The alternative I suppose is to use setters instead.

**Review by mrlumbu — APPROVED** [review]
